### PR TITLE
(GH-65) Managed identity access token not being cached in hiera lookups

### LINKED ISSUE #65: Access token never cached
Puppet apply fails with the following error:

```
Could not retrieve catalog from remote server: Error 500 on SERVER: Server Error: Puppet::Parser::Compiler failed with error RuntimeError: {"error":"invalid_request","error_description":"Temporarily throttled, too many requests"} on node XXXX
```

This is caused as the `access_token` is never cached inside `lookup.rb`, causing this module to hit the rate limit for the instance metadata service.

### Environment
**Version:**
Module version `1.1.0`
Puppet Enterprise v2019.8.5

**`hiera.yaml` contents:**
```
---
version: 5

defaults:
  datadir: "data"

hierarchy:
  - name: "Yaml backend"
    data_hash: yaml_data
    paths:
      - "nodes/%{trusted.certname}.yaml"
      - "common.yaml"
  - name: 'Azure Key Vault Secrets'
    lookup_key: azure_key_vault::lookup
    options:
      vault_name: casual-finch-keyvault
      vault_api_version: '2016-10-01'
      metadata_api_version: '2018-02-01'
```

### Troubleshooting

Looking at the Puppet server log we can see the corresponding log entries indicating this is failing in the `lookup.rb` file:
```
2021-04-18T15:57:45.856Z ERROR [qtp852575914-224] [puppetserver] Puppet Puppet::Parser::Compiler failed with error RuntimeError: {"error":"invalid_request","error_description":"Temporarily throttled, too many requests"} on node XXXX
2021-04-18T15:57:45.858Z ERROR [qtp852575914-224] [puppetserver] Puppet Server Error: Puppet::Parser::Compiler failed with error RuntimeError: {"error":"invalid_request","error_description":"Temporarily throttled, too many requests"} on
 node XXXX
/etc/puppetlabs/code/environments/production/modules/azure_key_vault/lib/puppet_x/tragiccode/azure.rb:15:in `get_access_token'
/etc/puppetlabs/code/environments/production/modules/azure_key_vault/lib/puppet/functions/azure_key_vault/lookup.rb:24:in `lookup_key'
```

Adding in some print debug code I was able to confirm that `get_access_token` was being called multiple times per puppet run:
```
[root@puppet-test tmp]# grep lookup_key /tmp/kvdebug | wc -l
42
[root@puppet-test tmp]# grep get_access_token /tmp/kvdebug | wc -l
41
```

### Commentary

This is the `lookup.rb` equivalent of #58, caused due to the Instance Metadata service only allowing 5 requests per second, as per Microsoft documentation.

> In general, requests to IMDS are limited to 5 requests per second. Requests exceeding this threshold will be rejected with 429 responses. Requests to the Managed Identity category are limited to 20 requests per second and 5 concurrent requests.

Source: https://docs.microsoft.com/en-us/azure/virtual-machines/windows/instance-metadata-service?tabs=windows#rate-limiting

### PR DESCRIPTION
**Pull request by dowlingw**
Modifies `lookup.rb` to cache managed identity access token to resolve #65

### CODE REVIEW

**Comment by TraGicCode** [comment]
Hey @dowlingw,

Thank you for the Pull Request.  Do you happen to know if a newer version of puppet changed/removed the methods i was originally using which caused the caching to be broken for you?

Thanks!

**Comment by dowlingw** [comment]
`Puppet::LookupContext` is still supported, but having a look through history it looks like the `context.cache()` was never implemented for the access token in `lookup.rb`, so from what I can see this has likely never been caching the access token.

File: lib/puppet/functions/azure_key_vault/lookup.rb [hunk 1]
```diff
@@ -11,11 +11,11 @@ def lookup_key(secret_name, options, context)
...
     return context.cached_value(secret_name) if context.cache_has_key(secret_name)
-    access_token = if context.cache_has_key('access_token')
-                     context.cached_value('access_token')
-                   else
-                     TragicCode::Azure.get_access_token(options['metadata_api_version'])
-                   end
+    access_token = context.cached_value('access_token')
+    if access_token.nil?
+      access_token = TragicCode::Azure.get_access_token(options['metadata_api_version'])
+      context.cache('access_token', access_token)
```
Comment (TraGicCode) [inline lib/puppet/functions/azure_key_vault/lookup.rb:17]: wow, i cannot believe i actually never cached the value!

**Review by TraGicCode — APPROVED** [review]

File: lib/puppet/functions/azure_key_vault/lookup.rb [hunk 2]
```diff
@@ -11,11 +11,11 @@ def lookup_key(secret_name, options, context)
     # This is a reserved key name in hiera
     return context.not_found if secret_name == 'lookup_options'
     return context.cached_value(secret_name) if context.cache_has_key(secret_name)
-    access_token = if context.cache_has_key('access_token')
-                     context.cached_value('access_token')
-                   else
-                     TragicCode::Azure.get_access_token(options['metadata_api_version'])
-                   end
+    access_token = context.cached_value('access_token')
+    if access_token.nil?
```
Comment (TraGicCode) [inline lib/puppet/functions/azure_key_vault/lookup.rb:15]: wow, i cannot believe i actually never cached the value!!!

Can you write a unit test (https://github.com/TraGicCode/tragiccode-azure_key_vault/blob/master/spec/functions/azure_key_vault_lookup_spec.rb) to assert that the value is stored in the cache when it's not returned from the cache?

**Review by TraGicCode — CHANGES_REQUESTED** [review]

**Comment by TraGicCode** [comment]
Hey @dowlingw ,

Ignore all comments above this one.  Thanks for fixing this bug! Can you complete the following?

1. Add a unit test to verify that the access token is cached when it's fetched because it was not found in the cache? ( Add this here https://github.com/TraGicCode/tragiccode-azure_key_vault/blob/master/spec/functions/azure_key_vault_lookup_spec.rb )

**Review by TraGicCode — CHANGES_REQUESTED** [review]
Hey @dowlingw ,

Ignore all comments above this one.  Thanks for fixing this bug! Can you complete the following?

1. Add a unit test to verify that the access token is cached when it's fetched because it was not found in the cache? ( Add this here https://github.com/TraGicCode/tragiccode-azure_key_vault/blob/master/spec/functions/azure_key_vault_lookup_spec.rb )

**Review by TraGicCode — APPROVED** [review]
LGTM.  Thank you for your contribution!

**Pull request merged**